### PR TITLE
Gradient loss logs

### PR DESCRIPTION
**Pull request by MatteoKartoon**
- Added gradient loss logs on wandb
- Modified the validation loss components logs: logging average over the whole validation set
- Modified the loss components logs: logging the average between the two devices
- Updated the readme file
- Added gradient loss to the total loss computation

### CODE REVIEW

File: scripts/train.py [hunk 1]
```diff
@@ -237,14 +246,16 @@ def iteration_over_batches_validation(self, epoch, info_progress=None, step_idx=
...
                        "Gradient Norm": self.last_grad_norm,
-                       "BCE loss validation": self.loss_components_validation['bce'],
-                       "SSIM loss validation": self.loss_components_validation['ssim'],
-                       "MAE loss validation": self.loss_components_validation['mae'],
-                       "IoU loss validation": self.loss_components_validation['iou'],
-                       "BCE loss training": self.loss_components_train['bce'],
-                       "SSIM loss training": self.loss_components_train['ssim'],
-                       "MAE loss training": self.loss_components_train['mae'],
-                       "IoU loss training": self.loss_components_train['iou']
+                       "BCE loss validation": loss_comp_val_aver_dev['bce'],
```
Comment (joelkartoon) [inline scripts/train.py:249]: move the training part into the train function (not validation)

File: scripts/train.py [hunk 2]
```diff
@@ -220,14 +220,23 @@ def _get_loss_key(self,epoch):
...
         #Loop over the batches
         total_loss=0
+        loss_comp_val_sum={k:0 for k in self.loss_components_train.keys()}
+        loss_comp_val_aver={k:0 for k in self.loss_components_train.keys()}
+        #Compute the sum of the losses over the validation set batches
         for batch_idx, batch in enumerate(self.validation_loader):
             self._batch(batch, batch_idx, epoch, validation=True)
             # Logger
             total_loss+=self.loss_dict_validation[self._get_loss_key(epoch)]
+            for k in self.loss_components_validation.keys():
```
Comment (joelkartoon) [inline scripts/train.py:230]:
```
# accumulate - double-check whether we sum or average across mini-batch
for loss_name, loss_value in self.loss_components_validation.items():
    loss_dict[loss_name] += loss_value

[..]

# average across batches
for loss_name, loss_accumulated_value in self.loss_dict.items():
    self.loss_dict = loss_accumulated_value / len([..])

# average across devices
for loss_name, loss_accumulated_value in self.loss_dict.items():
    self.loss_dict = self.average_between_devices([..])
```

File: scripts/train.py [hunk 3]
```diff
@@ -220,33 +220,45 @@ def _get_loss_key(self,epoch):
...
         for batch_idx, batch in enumerate(self.validation_loader):
             self._batch(batch, batch_idx, epoch, validation=True)
             # Logger
             total_loss+=self.loss_dict_validation[self._get_loss_key(epoch)]
+            for loss_name, loss_value in self.loss_components_validation.items():
+                loss_components_dict[loss_name]+=loss_value
+
         #Compute the average of the losses over the validation set
+        loss_components_dict={loss_name:loss_accumulated_value/len(self.validation_loader) for loss_name, loss_accumulated_value in loss_components_dict.items()}
         average_loss=total_loss/len(self.validation_loader)
```
Comment (joelkartoon) [inline scripts/train.py:235]: can we simplify further and remove `total_loss`
and have something like

```
average_loss = sum(loss_components_dict.values())
```

**Review by joelkartoon — APPROVED** [review]
